mne_bids/write.py

```diff
@@ -1372,3 +1382,109 @@ def write_anat(bids_root, subject, t1w, session=None, acquisition=None,
+
+    Parameters
+    ----------
+    bids_basename : str | BIDSPath
+        The base filename of the BIDS compatible files. Typically, this can be
+        generated using :func:`mne_bids.make_bids_basename`. The basename
+        requires at least the subject entity to work, and
+        needs to uniquely identify the scan inside the scans.tsv file.
+    bids_root : str | pathlib.Path
+        Path to the root of the BIDS directory.
+    verbose : bool
+        If verbose is True, this will print which files were removed.
+    """
+    if isinstance(bids_basename, str):
+        params = _parse_bids_filename(bids_basename, verbose)
+        bids_basename = BIDSPath(subject=params.get('sub'),
+                                 session=params.get('ses'),
+                                 recording=params.get('rec'),
+                                 acquisition=params.get('acq'),
+                                 processing=params.get('proc'),
+                                 space=params.get('space'),
+                                 run=params.get('run'),
+                                 task=params.get('task'))
+
+    if bids_basename.subject is None:
+        raise RuntimeError(f'Deleting a scan requires the bids_basename '
+                           f'to have at least subject defined. '
+                           f'You passed in {bids_basename}')
+
+    scans_fpath = _find_matching_sidecar(bids_basename, bids_root,
+                                         suffix='scans.tsv',
+                                         allow_fail=False)
+    scans_tsv = _from_tsv(scans_fpath)
+    scans_fnames = scans_tsv['filename']
+
+    # find the full filename to delete
+    matching_basenames = [fname for fname in scans_fnames
+                          if str(bids_basename) in fname]
+    if len(matching_basenames) > 1:
+        raise RuntimeError(f'Deleting scan requires a unique bids_basename '
+                           f'to parse in the scans.tsv file. '
+                           f'Found more than 1 ({matching_basenames})...')
+    elif len(matching_basenames) == 0:
+        raise RuntimeError(f'Trying to delete {bids_basename} '
+                           f'from scans.tsv, but no files were found...')
+    else:
+        bids_fname = matching_basenames[0]
+
+    # delete the file and its sidecar files
+    parent_path = Path(op.dirname(scans_fpath))
+    scans_fpaths = parent_path.rglob(f'*{bids_basename}*')
+    for fpath in scans_fpaths:
+        os.remove(fpath)
+
+    if [fpath for fpath in parent_path.rglob('*')] == []:
+        # remove the entire subject if there is nothing there
+        delete_participant(bids_basename.subject,
+                           bids_root=bids_root,
+                           verbose=verbose)
+    else:
+        # resave the scans.tsv file
+        scans_tsv = _drop(scans_tsv, bids_fname, 'filename')
+        _to_tsv(scans_tsv, scans_fpath)
+
+
+def delete_participant(subject, bids_root, verbose=True):
```

Comment:
I'm a bit uncertain of an API like this. In the future will you also need a `delete_session` function?

Couldn't we instead have some checks? E.g., remove all the files that match a certain regular expression/wildcard/keyword argument pair (subject='blah', run='blah' etc.) ? Then finally sanitize the directories + scans.tsv + participants.tsv.

Reply:
This would be possible now thanks to `BIDSPath.match()`.